package/network/services/dropbear/files/dropbear.init

```diff
@@ -196,6 +260,7 @@ boot()
 start_service()
 {
 	hk_generate_as_needed
+	adjust_permissions
```

Comment:
That fixes up stuff on the target. On every start. With precious writes to the rootfs.
That should and can be avoided.

First idea:
Don't `touch` the *_host_key files in the install step, instead commit empty ones in ./files and use `INSTALL_CONF` (0600). That will allow users to replace the empty ones with theirs in the image and they always have the correct permissions without the need to fix anything up afterwards.

Reply:
My point of view is that we should (!) do this every service start because some targets have only ssh access to them and if dropbear can't start because of broken keys and/or incorrect ownership/permissions - we should avoid it at all costs.

I would like to listen @M95D: how do you bootstrapping authorized keys/host keys for dropbear in your own setup? Details, please. :)

Reply:
By that logic we'd have to fixup the whole of /rom with on-target post process scripts. That's backwards imho, but we can ensure that the /rom contents are properly set up so users don't run into these sort of issues.

And I don't bundle my keys in the image. I `sysupgrade -u` so they're kept.

If you want to bundle keys in the image it's best to use supported ways. There's currently none for dropbear keys I think? My idea above would create a simple one.

Reply:
My setup has serial access. I used that to fix file permissions. I still have this image installed in my router. These are file permission from /rom:
```
root@GRAPHRT:/rom/etc/dropbear# ls -la
drwxrwxrwx    2 root     root           131 Jun 30 16:03 ./
drwxrwxrwx   37 root     root          1547 Jun 30 16:03 ../
-rwxrw-rw-    1 root     root           389 Jun 30 16:03 authorized_keys*
-rwxrw-rw-    1 root     root           141 Jun 30 16:03 dropbear_ecdsa_host_key*
-rwxrw-rw-    1 root     root            83 Jun 30 16:03 dropbear_ed25519_host_key*
-rwxrw-rw-    1 root     root           805 Jun 30 16:03 dropbear_rsa_host_key*
```
After installing this, I changed the file permissions on the files to be included in the squashfs and the next image was built with the proper permissions (if I recall corectly). I can't test it now.

To include keys (and lots of other files) in the squashfs image, I put them in **files/** subdir before building the image. On my system, **files** is a symlink to a NFS share on the router, where I sync with the files that are in use on that router.

I think that file permissions are kept unchanged when squashfs is generated. When I saw the problem, I changed the permisissions on the source files before the next build and the new image had the correct permissions.

Setting file permissions on every boot would be VERY wrong. I think that setting file permission from the Makefile, when OpenWrt is buit is enough.

Reply:
To be clear: I'm doing ownership/permissions fixes at least two times - while building package/ROM (to ensure that if user supplied own files - they will have appropriate permissions otherwise user is hit by #10849) and on every service start (to ensure that every file on target was set up properly - contents, ownership, permissions). I don't want anyone to be stuck with working but unmanageable device (good ol' NOC's days). :)

Reply:
@dhewg If you have any ideas how to deal with host keys during ROM build and during service startup - write some code/snippets, please. :)

Reply:
@M95D I agree that setting ownership/permissions without conditions is somewhat bad idea - I'd fix that a bit later, stay tuned. =)

Reply:
I didn't test it, but this should work, no?

> Don't `touch` the *_host_key files in the install step, instead commit empty ones in ./files and use `INSTALL_CONF` (0600).

That's almost half a code snipped :)

Reply:
At boot, check permissions and don't write unless they are wrong?

Reply:
I'm stuck at permissions check in init script because there's no `stat` binary (busybox gadget) by default. Any ideas? :)

Reply:
We always may use "`ls -Lln "$1" | tr -s '\t ' ' '`" as fallback.